### PR TITLE
Snapshot: Rollback after restart.

### PR DESCRIPTION
**Pull request by OhmSpectator**
This pull request introduces significant changes to handling volume snapshots in the system with the primary goal of enabling rollback functionality after a system reboot without requiring persistent subscriptions. Additionally, serialization and deserialization have been implemented in a way that allows the serialized data to survive EVE upgrades, even if the data format undergoes changes.

The key changes in this PR include:

Reworked snapshot configuration handlers in volumemgr.go to improve their functionality and code organization. The new handlers (handleVolumesSnapshotConfigCreate, handleVolumesSnapshotConfigModify, and handleVolumesSnapshotConfigDelete) now call a common implementation function (handleVolumesSnapshotConfigImpl).

Significant rework of functions in handlesnapshot.go such as handleVolumesSnapshotCreate, handleVolumesSnapshotModify, and handleVolumesSnapshotDelete. These functions now call handleVolumesSnapshotConfigImpl with the appropriate action set in the snapshot config. The new functions (createSnapshot, rollbackSnapshot, and deleteSnapshot) have been introduced to handle the respective actions.

Updates to existing functions (publishVolumesSnapshotStatus, unpublishVolumesSnapshotStatus, and lookupVolumesSnapshotStatus) to improve their functionality and handle snapshot status more effectively. A new function (deserializeVolumesSnapshotStatus) has been introduced to read the snapshot status from a file, allowing rollback after system reboot.

Improvements in the rollback process within the Zedmanager's updatestatus module. Instead of restoring VolumeRefStatuses and AppInstanceConfig from a snapshot, the function now creates or updates a VolumesSnapshotConfig to indicate a rollback action. This change simplifies the rollback process and makes it more resilient to system reboots.

Various code refactoring, cleanup, and optimization changes throughout the codebase to improve clarity, readability, and maintainability.

### CODE REVIEW

**Comment by OhmSpectator** [comment]
Pushed changes to the Yetus issues.

**Comment by OhmSpectator** [comment]
Pushed another fix for the Yetus issue

**Comment by OhmSpectator** [comment]
Rebased on master

File: pkg/pillar/cmd/volumemgr/handlesnapshot.go
Comment (milan-zededa) [inline pkg/pillar/cmd/volumemgr/handlesnapshot.go:124]: Perhaps this whole part that checks if json files contains required fields and logs warnings about unexpected fields could be generalized into a helper function? I see that you have it twice here and maybe it could be also used for some other current/future persisted jsons.

Comment (OhmSpectator) [reply pkg/pillar/cmd/volumemgr/handlesnapshot.go:124]: Yes, I thought about it. I can rework it later, especially if you recommend where to put such a function (do we have some generic helpers? `lib/utils`?)
But I would do it later, not blocking the merge of the current PR, if you don't mind.

File: pkg/pillar/types/volumetypes.go
Comment (eriknordmark) [inline pkg/pillar/types/volumetypes.go:304]: Since the json Encode omits fields which have their default values it might make sense to add to this comment that such fields must have a non-default value?

File: pkg/pillar/utils/generics.go
Comment (eriknordmark) [inline pkg/pillar/utils/generics.go:152]: I realize you don't need this nesting now, but when we do I think the field names can conflict.

If you have e.g.,
type top struct {
    a int
    b nest
}
type next struct {
   a int
}

Then field.Name will be "a" for both top.a and top.nest.a
One could fix this by adding a prefix argument to ExtractFields so be able to e.g., construct these dotten names I used as an example.
Or just handle the top level fields here??

Some other alternatives:
For the critical fields one can do something simpler by checkig that their values are not default aka empty. But that doesn't allow warning for extraneous fields. One can call json.Decode a second time setting the flag (I forget exactly where that needs to be set) to error on unknown fields, but I don't know if that will return an error string indocating the names of the extraneous fields, or just the first extra field.

File: pkg/pillar/cmd/volumemgr/handlesnapshot.go
Comment (eriknordmark) [reply pkg/pillar/cmd/volumemgr/handlesnapshot.go:124]: Might make sense to have this error/warning be a separate function in this file - returning an error when you need to bail and otherwise log warnings. That would make the deserialize function itself smaller.

File: pkg/pillar/utils/generics.go
Comment (OhmSpectator) [reply pkg/pillar/utils/generics.go:152]: Just in case, here is a link for a snippet that helps understand how it works:
https://go.dev/play/p/g7n6mn8nFnA

**Review by eriknordmark — COMMENTED** [review]
With these changes are you still creating a VolumeStatus before there is a VolumeConfig from the controller/zedagent?

If not, can we restore the log.Fatal in handleDeferredVolumeCreate()?

**Comment by OhmSpectator** [comment]
> With these changes are you still creating a VolumeStatus before there is a VolumeConfig from the controller/zedagent?
> 
> If not, can we restore the log.Fatal in handleDeferredVolumeCreate()?

As far as I remember, nope. I removed the persistent flag from the VolumeStatus topic and I do not serialize/deserialize VolumeStatus anyhow.

I'll check the ` handleDeferredVolumeCreate()`, ok.

File: pkg/pillar/utils/generics.go
Comment (eriknordmark) [reply pkg/pillar/utils/generics.go:152]: Maybe I misunderstood anynomous to mean 'unexported'. If it is instead what we have with a type without a name, then what you have is perfectly fine.

Comment (OhmSpectator) [reply pkg/pillar/utils/generics.go:152]: Yep, a field with a type but without a name is an anonymous field (or an embedded field) in terms of Go =)

File: pkg/pillar/cmd/zedmanager/zedmanager.go
Comment (eriknordmark) [inline pkg/pillar/cmd/zedmanager/zedmanager.go:223]: Somewhere we should document what these are about (basically tracking the AppInstanceConfig which we have just rolled back to until the controller/zedagent sends us that as AppInstanceConfig) and that they take precedence over the AppInstanceConfig received from the controller/zedagent. 

I don't know if this is the best place to do that.

Note that we also have pkg/pillar/docs/zedmanager.md (I hope), and we can use that to add high-level descriptions about snapshot and rollback.

File: pkg/pillar/cmd/zedmanager/handlevolumemgr.go
Comment (eriknordmark) [inline pkg/pillar/cmd/zedmanager/handlevolumemgr.go:312]: If this returns nil it would be better to have a log.Error or even log.Fatal than the deference of nil below.

File: pkg/pillar/cmd/zedmanager/zedmanager.go
Comment (eriknordmark) [inline pkg/pillar/cmd/zedmanager/zedmanager.go:682]: Might make sense to document that is islocal is set we give preference to the local config from a snapshot.

File: pkg/pillar/cmd/zedmanager/zedmanager.go
Comment (eriknordmark) [inline pkg/pillar/cmd/zedmanager/zedmanager.go:705]: Doesn't match function name.

**Review by eriknordmark — COMMENTED** [review]
Looks very good.

With these changes are you still creating a VolumeStatus before there is a VolumeConfig from the controller/zedagent?

If not, can we restore the log.Fatal in handleDeferredVolumeCreate()?

I think there is a small piece missing (cleanup any leftovers in /persist/snapshots after boot) but since this PR is quite large let's do that as a small follow-on.
(Such things can happen when an app instance and associated snapshots are deleted on the controller when the device is e.g. powered off, and we want to garbage collect such disk space.)

**Comment by OhmSpectator** [comment]
> I think there is a small piece missing (cleanup any leftovers in /persist/snapshots after boot) but since this PR is quite large let's do that as a small follow-on.

There is an issue created for the cleanups (EV-606). Actually, there is one more case to be fixed. But yep, I want to address these issues in the upcoming PRs.

**Comment by OhmSpectator** [comment]
Rebased on master

**Comment by OhmSpectator** [comment]
> With these changes are you still creating a VolumeStatus before there is a VolumeConfig from the controller/zedagent?
> 
> If not, can we restore the log.Fatal in handleDeferredVolumeCreate()?

I'll test it now.

**Comment by OhmSpectator** [comment]
Restored the log.Fatal in handleDeferredVolumeCreate()

File: pkg/pillar/cmd/zedmanager/zedmanager.go
Comment (OhmSpectator) [reply pkg/pillar/cmd/zedmanager/zedmanager.go:705]: Fixed

File: pkg/pillar/cmd/zedmanager/zedmanager.go
Comment (OhmSpectator) [reply pkg/pillar/cmd/zedmanager/zedmanager.go:685]: Done

File: pkg/pillar/cmd/zedmanager/zedmanager.go
Comment (OhmSpectator) [reply pkg/pillar/cmd/zedmanager/zedmanager.go:452]: Done

File: pkg/pillar/cmd/zedmanager/handlevolumemgr.go
Comment (OhmSpectator) [reply pkg/pillar/cmd/zedmanager/handlevolumemgr.go:312]: Added a log.Fatal

File: pkg/pillar/cmd/zedmanager/zedmanager.go
Comment (OhmSpectator) [reply pkg/pillar/cmd/zedmanager/zedmanager.go:223]: We will rework the rolling back mechanism soon so that the controller sends us the target (snapshotted) config immediately with the rollback command. It may change the logic of this subscription. So, I would prefer to document it after we finish the rollback command's final version.

**Comment by OhmSpectator** [comment]
I will also try refactoring the JSON marshaling/unmarshaling code to address the request to generalize this part.

**Comment by OhmSpectator** [comment]
@eriknordmark, @milan-zededa, I hope I've addressed all the comments now.

File: pkg/pillar/utils/deserialize.go
Comment (milan-zededa) [inline pkg/pillar/utils/deserialize.go:90]: More common signature for deserialize/unmarshal functions in Go is to receive pointer to empty structure and perform all the reflect magic inside (so that caller does not have to deal with reflect). For example, [json.Unmarshal](https://pkg.go.dev/encoding/json#Unmarshal) takes `any` and expect it to point to empty struct that json data should be unmarshalled into.
With custom tags marking critical fields the function could be declared as:
```
func DeserializeToStruct(filename string, structure any) error
```

File: pkg/pillar/utils/deserialize.go
Comment (milan-zededa) [inline pkg/pillar/utils/deserialize.go:40]: What is the point of this function, seems it just forwards the call to `json.Unmarshal`.

**Review by milan-zededa — APPROVED** [review]
Thanks for moving struct deserialization to a standalone utility.
Added just few comments for minor improvements.
Let's start eden tests in the meantime.

File: pkg/pillar/types/volumetypes.go
Comment (OhmSpectator) [reply pkg/pillar/types/volumetypes.go:288]: Sounds interesting. I'd like to check this tag approach.

File: pkg/pillar/utils/deserialize.go
Comment (OhmSpectator) [reply pkg/pillar/utils/deserialize.go:90]: Yeah, I also don't like this `reflect.TypeOf` in the caller.
I'll try to refactor this.

File: pkg/pillar/cmd/volumemgr/handlesnapshot.go
Comment (OhmSpectator) [reply pkg/pillar/cmd/volumemgr/handlesnapshot.go:186]: Yeah, it looks much cleaner, thanks! Will change.

File: pkg/pillar/utils/deserialize.go
Comment (OhmSpectator) [reply pkg/pillar/utils/deserialize.go]: Done

File: pkg/pillar/types/volumetypes.go
Comment (OhmSpectator) [reply pkg/pillar/types/volumetypes.go:288]: Done

File: pkg/pillar/cmd/volumemgr/handlesnapshot.go
Comment (OhmSpectator) [reply pkg/pillar/cmd/volumemgr/handlesnapshot.go]: Done

File: pkg/pillar/utils/deserialize.go
Comment (OhmSpectator) [reply pkg/pillar/utils/deserialize.go]: Done

File: pkg/pillar/utils/deserialize.go
Comment (milan-zededa) [inline pkg/pillar/utils/deserialize.go:61]: Maps are passed by reference, you do not need to pass them with pointer.

File: pkg/pillar/types/volumetypes.go
Comment (milan-zededa) [reply pkg/pillar/types/volumetypes.go:288]: Thanks

File: pkg/pillar/utils/deserialize.go
Comment (milan-zededa) [reply pkg/pillar/utils/deserialize.go]: Thanks

File: pkg/pillar/types/volumetypes.go
Comment (milan-zededa) [reply pkg/pillar/types/volumetypes.go:288]: Looks nice

File: pkg/pillar/utils/deserialize.go
Comment (OhmSpectator) [reply pkg/pillar/utils/deserialize.go:61]: Thanks! I will fix it!

Comment (OhmSpectator) [reply pkg/pillar/utils/deserialize.go]: Done

**Review by milan-zededa — APPROVED** [review]
Restarting tests

**Comment by OhmSpectator** [comment]
I don't know how to interpret the tests failures...

File: pkg/pillar/types/volumetypes.go
Comment (eriknordmark) [reply pkg/pillar/types/volumetypes.go]: Sorry my confusion with protobuf encoding.

File: pkg/pillar/cmd/zedmanager/zedmanager.go
Comment (eriknordmark) [reply pkg/pillar/cmd/zedmanager/zedmanager.go:223]: OK will you do that rework as part of this PR or as a follow-on PR?

**Comment by eriknordmark** [comment]
> I don't know how to interpret the tests failures...

Which test failures? I don't see any in the current checks.

File: pkg/pillar/utils/deserialize.go
Comment (eriknordmark) [inline pkg/pillar/utils/deserialize.go:81]: Nit, but it wouldn't hurt to comment that this will check and fail if any fields with mandatory=true are not present.

File: pkg/pillar/cmd/zedmanager/handlevolumemgr.go
Comment (eriknordmark) [inline pkg/pillar/cmd/zedmanager/handlevolumemgr.go:327]: Why don't you propagate the warnings? Is due to the warning you generate in createSnapshot when the snapshot already exists? If so, please add a comment about this.

File: pkg/pillar/cmd/zedmanager/zedmanager.go
Comment (eriknordmark) [inline pkg/pillar/cmd/zedmanager/zedmanager.go:480]: Shouldn't there be a log.Warning here? This should never happen AFAICT.

**Review by eriknordmark — APPROVED** [review]
Looks very nice; I didn't realize one can add custom tags to structs as you did with the man datory tag.

File: pkg/pillar/cmd/zedmanager/zedmanager.go
Comment (OhmSpectator) [reply pkg/pillar/cmd/zedmanager/zedmanager.go:223]: It will be part of a new PR.
We need this version to be merged to have at least the initial version of the snapshot feature working in the master branch. After that, we will start with the rework on Cloud-side and then - EVE.

File: pkg/pillar/utils/deserialize.go
Comment (OhmSpectator) [reply pkg/pillar/utils/deserialize.go:83]: Done. 
I also added a check that the pointer passed to the function is a pointer to a struct.

File: pkg/pillar/cmd/zedmanager/zedmanager.go
Comment (OhmSpectator) [reply pkg/pillar/cmd/zedmanager/zedmanager.go:480]: No, it's expected behavior. On the system boot, after restart EVE goes through the list of apps, and for each of them, it looks for available snapshots serialized. If the snapshot parsed is not for this particular app - EVE skips it and goes further.

**Comment by OhmSpectator** [comment]
> Which test failures? The "Unchanged files with check annotations" from the Eden run shown in the annotated diffs?

On the PR build, for example, as I mentioned before.
```
Error: error building "lfedge/eve-newlog:f5f94fbc8abb8cecae0718fea66d6bf3b88d1e95": descriptor for platform {amd64 linux  [] } has no information on the platform: &v1.Descriptor{MediaType:"application/vnd.oci.image.index.v1+json", Size:437, Digest:v1.Hash{Algorithm:"sha256", Hex:"3f5b4fea8b57395932f4bd16f89d6bb00288ea7c37f5523a089398f54f356097"}, Data:[]uint8(nil), URLs:[]string(nil), Annotations:map[string]string{"org.opencontainers.image.ref.name":"docker.io/lfedge/eve-newlog:f5f94fbc8abb8cecae0718fea66d6bf3b88d1e95"}, Platform:(*v1.Platform)(nil)}
```
I don't think this is relevant...

File: pkg/pillar/cmd/zedmanager/zedmanager.go
Comment (eriknordmark) [reply pkg/pillar/cmd/zedmanager/zedmanager.go:480]: OK

**Comment by OhmSpectator** [comment]
Idk, why the test runs take that long in a queue now. Meanwhile, I was able to run a full ztest set (run 39945). The PR passed it.

**Comment by OhmSpectator** [comment]
Added a unit test for the deserialization function.

**Comment by eriknordmark** [comment]
> @eriknordmark, I'm not sure why the test runs are taking so long in the queue lately. However, I wanted to let you know that I was able to successfully run the entire ztest set (run 39945) and the PR passed.

Thanks. I'll merge this once the yetus+build+unit test passes on the latest bits.

I sent an email to the lf-edge eve mailing list about the queue of workflows and needing to manually cancel tons of them.

**Pull request merged**
